File: patch.py
import re

with open('old_google_apps_script_utf8.js', 'r', encoding='utf-8') as f:
    old_code = f.read()

with open('google-apps-script.js', 'r', encoding='utf-8') as f:
    new_code = f.read()

def extract(code, start_marker, end_marker):
    start = code.find(start_marker)
    if start == -1: return ''
    end = code.find(end_marker, start)
    if end == -1: end = len(code)
    return code[start:end]

submit_old = extract(old_code, '// ACTION: SUBMIT_REGISTRATION', '// ACTION: CHECKIN')
reset_old = extract(old_code, '// ACTION: RESET_REGISTRATIONS', '// ACTION: SAVE_REG_CONFIG')
get_reg_old = extract(old_code, 'if (action === "get_registration") {', 'if (action === "get_shift_registrations") {')
get_shift_reg_old = extract(old_code, 'if (action === "get_shift_registrations") {', 'if (action === "sync_roster") {')

submit_new = extract(new_code, '// ACTION: SUBMIT_REGISTRATION', '// ACTION: CHECKIN')
reset_new = extract(new_code, '// ACTION: RESET_REGISTRATIONS', '// ACTION: SAVE_REG_CONFIG')
get_reg_new = extract(new_code, 'if (action === "get_registration") {', 'if (action === "get_reg_config") {')

end_marker_new = 'if (action === "sync_roster") {' if new_code.find('if (action === "sync_roster") {') != -1 else '// Pre-flight check'
get_shift_reg_new = extract(new_code, 'if (action === "get_shift_registrations") {', end_marker_new)

new_code = new_code.replace(submit_new, submit_old)
new_code = new_code.replace(reset_new, reset_old)
new_code = new_code.replace(get_reg_new, get_reg_old)
new_code = new_code.replace(get_shift_reg_new, get_shift_reg_old)

with open('google-apps-script.js', 'w', encoding='utf-8') as f:
    f.write(new_code)

print('Success Python patch')

File: patch.js
const fs = require('fs');
const oldCode = fs.readFileSync('old_google_apps_script_utf8.js', 'utf8');
const newCode = fs.readFileSync('google-apps-script.js', 'utf8');

function extractBlock(code, startMarker, endMarker) {
    const startIndex = code.indexOf(startMarker);
    if (startIndex === -1) return null;
    let endIndex = code.indexOf(endMarker, startIndex);
    if (endIndex === -1) endIndex = code.length;
    return code.substring(startIndex, endIndex);
}

const submitRegOld = extractBlock(oldCode, '// ACTION: SUBMIT_REGISTRATION', '// ACTION: CHECKIN');
const resetRegOld = extractBlock(oldCode, '// ACTION: RESET_REGISTRATIONS', '// ACTION: SAVE_REG_CONFIG');
const getRegOld = extractBlock(oldCode, 'if (action === "get_registration") {', 'if (action === "get_shift_registrations") {');
const getShiftRegOld = extractBlock(oldCode, 'if (action === "get_shift_registrations") {', '//==========================================');

let modifiedCode = newCode;

function replaceBlock(code, startMarker, endMarker, newBlock) {
    const startIndex = code.indexOf(startMarker);
    const endIndex = code.indexOf(endMarker, startIndex);
    if (startIndex !== -1 && endIndex !== -1 && newBlock) {
        return code.substring(0, startIndex) + newBlock + code.substring(endIndex);
    }
    return code;
}

modifiedCode = replaceBlock(modifiedCode, '// ACTION: SUBMIT_REGISTRATION', '// ACTION: CHECKIN', submitRegOld);
modifiedCode = replaceBlock(modifiedCode, '// ACTION: RESET_REGISTRATIONS', '// ACTION: SAVE_REG_CONFIG', resetRegOld);
modifiedCode = replaceBlock(modifiedCode, 'if (action === "get_registration") {', 'if (action === "get_shift_registrations") {', getRegOld);
modifiedCode = replaceBlock(modifiedCode, 'if (action === "get_shift_registrations") {', '//==========================================', getShiftRegOld);

fs.writeFileSync('google-apps-script.js', modifiedCode);
console.log('Restored the 4 blocks!');
